### ISSUE TITLE
Confirm CCD pixel coords - edge or center?

### ISSUE THREAD

**Issue by taldcroft** [sot/starcheck#273]
In obsid 20903 a proseco-generated catalog gives a Angle too large warning for a guide star (using starcheck 12.1).  It looks like current master would give the same, though I haven't checked if things are changed in any open proseco-related PRs here.  See also:
https://github.com/sot/starcheck/blob/fe36990463efafbe1af694eb5f54be9bb4e1cc83/starcheck/src/lib/Ska/Starcheck/Obsid.pm#L1096
```
In [10]: aca.guides
Out[10]: 
<GuideTable length=5>
    id        ra        dec       yang     zang     row     col   ...  maxmag p_acq  dim   res  halfw  sz 
  int32    float64    float64   float64  float64  float64 float64 ... float32 int64 int64 int64 int64 str3
--------- ---------- ---------- -------- -------- ------- ------- ... ------- ----- ----- ----- ----- ----
734532120 235.291877 -11.868950 -2292.77  2460.61  471.08  502.19 ...    9.28 0.000     1     1    25  6x6
734660656 236.504683 -11.750419  -463.74 -1425.58   98.61 -281.28 ...   10.85 0.000     1     1    25  6x6
734663064 236.031515 -12.061874 -2075.51  -226.63  423.28  -41.35 ...   10.95 0.000     1     1    25  6x6
734139376 236.446474 -10.971253  2114.05  -300.71 -419.40  -54.88 ...   11.13 0.000     1     1    25  6x6
734662376 236.684476 -11.553129   416.53 -1788.18  -78.29 -354.18 ...   11.18 0.000     1     1    25  6x6

In [11]: 511.5 - (7 + 1 + 8.0 / 5)  # Starcheck test NOT OK
Out[11]: 501.9

In [12]: 512.0 - (7 + 1 + 8.0 / 5)  # Proseco limit OK
Out[12]: 502.4
```

**Comment by taldcroft** [comment]
But now I'm confused with 48510, which has an angle row too large but the math doesn't make sense to me.
```
In [15]: aca = get_aca_catalog(48510, dither=(8,8))

In [17]: aca.guides.get_id(248268384)['row']
Out[17]: -494.42184017358659

In [18]: -512.5 + (7 + 8 + 8/5)  # Starcheck should be OK?
Out[18]: -495.9

In [19]: -512.0 + (7 + 8 + 8/5)  # Proseco OK
Out[19]: -495.4
```

**Comment by jeanconn** [comment]
Regarding "It looks like current master would give the same," did you run and get the same?  I'm not there yet, but I thought I implement about what we had in mind.

**Comment by taldcroft** [comment]
Current master has the same row/col min/max values and the same yag/zag to row/col transform as 12.1 so I was expecting to get the same answer.  But indeed I did not actually run it.

**Comment by jeanconn** [comment]
That's fine.  I couldn't remember if I trimmed any of the guide padding.  Probably the changes in that made it into master were all about the acq boxes off the CCD and tracking the two axes dither.

**Comment by taldcroft** [comment]
I did some more digging and here is what I believe:

- The perl `toPixels` (from `Ska::ACACoordConvert`) uses the standard "ground" polynomial coefficients, same as `chandra_aca.transform.yagzag_to_pixels` (and vice-versa).  The numerical values are identical.
- The "ground" coefficients use the "edge" coordinate convention for row/col, just like flight.
- Proseco uses "edge" coordinates and ground coefficients (i.e. the defaults).
- The physical edges of the CCD in "edge" coordinates are -512.0 to +512.0.

If this is all correct it leads to a conclusion that starcheck has a bug and the row/col edge limit checking is off by 0.5 pixels.

**Comment by jeanconn** [comment]
> I did some more digging and here is what I believe:
> 
> The perl toPixels (from Ska::ACACoordConvert) uses the standard "ground" polynomial coefficients, same as chandra_aca.transform.yagzag_to_pixels (and vice-versa). The numerical values are identical.

Yes.  I'd just done the same confirmation of the coefficients (as it seemed to be what you were asking to do in the thread) and agree that toPixels is in "edge".

But toPixels also says 
```  
croak "Coordinate off of CCD" if $y > 511.5 || $y < -512.5 || $z > 511.5 || $z < -512.5;
```

**Comment by taldcroft** [comment]
Just 4 pixels of window pad is definitely too small since (if I'm not misreading the code) that window pad includes the 4 pixel halfwidth readout box size.  So that would be zero pixels of margin.

**Comment by jeanconn** [comment]
Yes, my understanding of "window pad" is that it is the 4 pixels of the readout window halfwidth.  I'm just not sure we need any pixel margin on guide stars.  That would probably be a separate study though.

**Comment by taldcroft** [comment]
Regarding first principles, a good start is getting a raw image (bright star, near the center of the CCD, and early in the mission), centroiding in row/col using edge coords and converting row/col to yag/zag using the flight coefficients.  This should match telemetered yag/zag.  That would establish definitively that the EEPROM coefficients are in edge coords.  

I've already established that the ground and flight coefficients are in the *same* coordinate system (which we believe to be edge).

**Comment by taldcroft** [comment]
YES, we need margin, since:
- There is always a pixel of slop in terms of where the readout window actually winds up.
-  True spacecraft attitude can have an error of few arcsec, but it could be more (don't have solid stats on this).  
- Gravity gradient / feed-forward torque disturbances can be significant.

Do you know what the SAUSAGE limit is?

**Comment by jeanconn** [comment]
I mean, I'd always prefer margin and I'm wondering if the guide algorithm should take that into account.

By "That would probably be a separate study though." I meant I'm not as fluent as you and @malgosias are (thanks to annie work) with what the PEA does at the edges.  Are the "row pad" and "col pad" that we have still commandable coordinates for the readout box (if so, there's at least 1 px of margin there).  And then if the next commanded box for readout will be off CCD does it immediately go to a SRCH state? (in which case half the time a delta between true spacecraft attitude and actual spacecraft attitude would work in our favor for keeping a box on the CCD).

> Gravity gradient / feed-forward torque disturbances can be significant.

Though I assume this would settle out as long as all of our stars weren't right at the edge.

Also, starcheck has a comment on the column pad "[because outer col is not full-sized]" which also has me more curious about those pixels.

> Do you know what the SAUSAGE limit is?

IIRC from our previous discussions about this, the issue with SAUSAGE is that it is doing belt and suspenders on this and is filtering in pixel and in yag/zag.

**Comment by jeanconn** [comment]
WRT sausage, I can't remember if we wrote this up more or talked about it more in correspondence?

SAUSAGE is using these limits
https://github.com/sot/fot_matlab_tools/blob/review_4.0/Characteristics/characteristics_Stars.m#L188

And these limits
https://github.com/sot/fot_matlab_tools/blob/review_4.0/Characteristics/characteristics_Stars.m#L200

And here the fixedErrorPad is the dither in pixels for guide star selection and dither + maneuver error for acq star selection

https://github.com/sot/fot_matlab_tools/blob/review_4.0/MissionScheduling/stars/StarSelector/%40StarSelector/private/checkOffChips.m#L13

but that edgeBuffer is 5 pixels per characteristics.

**Comment by taldcroft** [comment]
See: row-col-zero-location.ipynb in the proseco validation www dir.
